primitives/io/src/batch_verifier.rs

```diff
@@ -0,0 +1,150 @@
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! This is part of the Substrate runtime.
+
+use sp_core::{ed25519, sr25519, crypto::Pair, traits::CloneableSpawn};
+use std::sync::{Arc, atomic::{AtomicBool, Ordering as AtomicOrdering}};
+use futures::{future::FutureExt, task::FutureObj, channel::oneshot};
+
+#[derive(Debug, Clone)]
+struct Ed25519BatchItem {
+	signature: ed25519::Signature,
+	pub_key: ed25519::Public,
+	message: Vec<u8>,
+}
+
+#[derive(Debug, Clone)]
+struct Sr25519BatchItem {
+	signature: sr25519::Signature,
+	pub_key: sr25519::Public,
+	message: Vec<u8>,
+}
+
+/// Batch verifier.
+///
+/// Used to parallel-verify signatures for runtime host. Provide task executor and
+/// just push (`push_ed25519`, `push_sr25519`) as many signature as you need. At the end,
+/// call `verify_and_clear to get a result. After that, batch verifier is ready for the
+/// next batching job.
+pub struct BatchVerifier {
+	scheduler: Box<dyn CloneableSpawn>,
+	sr25519_items: Vec<Sr25519BatchItem>,
+	invalid: Arc<AtomicBool>,
+	pending_tasks: Vec<oneshot::Receiver<()>>,
+}
+
+impl BatchVerifier {
+	pub fn new(scheduler: Box<dyn CloneableSpawn>) -> Self {
+		BatchVerifier {
+			scheduler,
+			sr25519_items: Default::default(),
+			invalid: Arc::new(false.into()),
+			pending_tasks: vec![],
+		}
+	}
+
+	fn spawn_verification_task(&mut self, f: impl FnOnce() -> bool + Send + 'static) {
+		// there is already invalid transaction encountered
+		if self.invalid.load(AtomicOrdering::Relaxed) { return; }
+
+		let invalid_clone = self.invalid.clone();
+		let (sender, receiver) = oneshot::channel();
+		self.pending_tasks.push(receiver);
+
+		self.scheduler.spawn_obj(FutureObj::new(async move {
+			if !f() {
+				invalid_clone.store(true, AtomicOrdering::Relaxed);
+			}
+			if let Err(_) = sender.send(()) {
+				// sanity
+				log::warn!("Verification halted while result was pendign");
+				invalid_clone.store(true, AtomicOrdering::Relaxed);
+			}
+		}.boxed())).expect("Scheduler should not fail");
+	}
+
+	/// Push ed25519 signature to verify.
+	pub fn push_ed25519(
+		&mut self,
+		signature: ed25519::Signature,
+		pub_key: ed25519::Public,
+		message: Vec<u8>,
+	) {
+		if self.invalid.load(AtomicOrdering::Relaxed) { return; } // there is already invalid transaction encountered
+
+		self.spawn_verification_task(move || ed25519::Pair::verify(&signature, &message, &pub_key));
+	}
+
+	/// Push sr25519 signature to verify.
+	pub fn push_sr25519(
+		&mut self,
+		signature: sr25519::Signature,
+		pub_key: sr25519::Public,
+		message: Vec<u8>,
+	) {
+		self.sr25519_items.push(Sr25519BatchItem { signature, pub_key, message });
```

Comment:
Seems rather doable to have a max batch size that trigger a spawn verification task every n sr25519 items.
But truely regarding the strategy to use it all depends on the workload of extrinsic processing, if it is heavy it could make sense to have such option to fail earlier.

Reply:
See benchmarks above, with 500 transactions in block this gives almost maximum performance gain (21%). And with less transactions in block we don't need this optimisation much.

Reply:
~4% for more or less full block

with batching using much less cpu